core/src/main/scala/cats/data/NonEmptyVector.scala

```diff
@@ -0,0 +1,230 @@
+package cats
+package data
+
+import scala.annotation.tailrec
+import scala.collection.immutable.VectorBuilder
+import scala.util.Try
+import cats.instances.vector._
+
+/**
+ * A data type which represents a non empty Vector.
+ */
+final case class NonEmptyVector[A] private (toVector: Vector[A]) {
+
+  /** Gets the element at the index, if it exists */
+  def get(i: Int): Option[A] =
+    Try(getUnsafe(i)).toOption
+
+  /** Gets the element at the index, or throws an exception if none exists */
+  def getUnsafe(i: Int): A = toVector(i)
+
+  /** Updates the element at the index, if it exists */
+  def updated(i: Int, a: A): Option[NonEmptyVector[A]] =
+    Try(updatedUnsafe(i, a)).toOption
+
+  /** Updates the element at the index, or throws an exeption if none exists */
+  def updatedUnsafe(i: Int, a: A):
+      NonEmptyVector[A] = NonEmptyVector(toVector.updated(i, a))
+
+  def head: A = toVector.head
+
+  def tail: Vector[A] = toVector.tail
+
+  /**
+   * remove elements not matching the predicate
+   */
+  def filter(f: A => Boolean): Vector[A] = toVector.filter(f)
+
+  /**
+   * Append another NonEmptyVector to this
+   */
+  def concat(other: NonEmptyVector[A]): NonEmptyVector[A] = NonEmptyVector(toVector ++ other.toVector)
+
+  /**
+    * Alias for concat
+    */
+  def ++(other: NonEmptyVector[A]): NonEmptyVector[A] = concat(other)
+
+  /**
+   * Append another Vector to this
+   */
+  def concat(other: Vector[A]): NonEmptyVector[A] = NonEmptyVector(toVector ++ other)
+
+  /**
+    * Alias for concat
+    */
+  def ++(other: Vector[A]): NonEmptyVector[A] = concat(other)
+
+  /**
+   * find the first element matching the predicate, if one exists
+   */
+  def find(f: A => Boolean): Option[A] = toVector.find(f)
+
+  /**
+   * Check whether at least one element satisfies the predicate.
+   */
+  def exists(f: A => Boolean): Boolean = toVector.exists(f)
+
+  /**
+   * Check whether all elements satisfy the predicate.
+   */
+  def forall(f: A => Boolean): Boolean = toVector.forall(f)
+
+  /**
+   * Left-associative fold using f.
+   */
+  def foldLeft[B](b: B)(f: (B, A) => B): B =
+    toVector.foldLeft(b)(f)
+
+  /**
+   * Right-associative fold using f.
+   */
+  def foldRight[B](lb: Eval[B])(f: (A, Eval[B]) => Eval[B]): Eval[B] =
+    Foldable[Vector].foldRight(toVector, lb)(f)
+
+  /**
+    * Applies f to all the elements
+    */
+  def map[B](f: A => B): NonEmptyVector[B] =
+    NonEmptyVector(toVector.map(f))
+
+  /**
+    *  Applies f to all elements and combines the result
+    */
+  def flatMap[B](f: A => NonEmptyVector[B]): NonEmptyVector[B] =
+    NonEmptyVector(toVector.flatMap(a => f(a).toVector))
+
+  /**
+    * Left-associative reduce using f.
+    */
+  def reduceLeft(f: (A, A) => A): A =
+    tail.foldLeft(head)(f)
+
+  /**
+    * Left-associative reduce using the Semigroup of A
```

Comment:
I think we should remove this comment: "Left-associative" since `Semigroup` is associative, it does not matter which order you do the combine, the result is the same.